### PR TITLE
[RDY] main: usage() improvements

### PR DESCRIPTION
**Pull request by ghost**
The diff between the actual usage() output might be hard to parse, so here's the diff (ignoring whitespace with `diff -w`) of `vim -h` output:

``` diff
--- oldout  2015-01-27 17:07:48.933613881 -0500
+++ newout  2015-01-27 17:07:31.913614073 -0500
@@ -1,9 +1,8 @@
-NVIM 0.0.0-alpha+201501262108 (compiled Jan 27 2015 17:07:42)
-
-usage: vim [arguments] [file ..]       edit specified file(s)
-   or: vim [arguments] -               read text from stdin
-   or: vim [arguments] -t tag          edit file where tag is defined
-   or: vim [arguments] -q [errorfile]  edit file with first error
+Usage:
+  nvim [arguments] [file ...]      Edit specified file(s)
+  nvim [arguments] -               Read text from stdin
+  nvim [arguments] -t <tag>        Edit file where tag is defined
+  nvim [arguments] -q [errorfile]  Edit file with first error

 Arguments:
    --          Only file names after this
@@ -18,17 +17,17 @@
    -M          Modifications in text not allowed
    -b          Binary mode
    -l          Lisp mode
-   -V[N][fname]        Be verbose [level N] [log messages to fname]
+  -V[N][file]           Be verbose [level N][log messages to file]
    -D          Debugging mode
    -n          No swap file, use memory only
    -r          List swap files and exit
-   -r (with file name) Recover crashed session
+  -r <file>             Recover crashed session
    -L          Same as -r
-   -A          start in Arabic mode
-   -H          Start in Hebrew mode
+  -A                    Start in Arabic mode
    -F          Start in Farsi mode
+  -H                    Start in Hebrew mode
    -T <terminal>   Set terminal type to <terminal>
-   -u <vimrc>      Use <vimrc> instead of any .vimrc
+  -u <nvimrc>           Use <nvimrc> instead of any .nvimrc
    --noplugin      Don't load plugin scripts
    -p[N]       Open N tab pages (default: one for each file)
    -o[N]       Open N windows (default: one for each file)
@@ -42,8 +41,8 @@
    -w <scriptout>  Append all typed commands to file <scriptout>
    -W <scriptout>  Write all typed commands to file <scriptout>
    --startuptime <file>    Write startup timing messages to <file>
-   -i <viminfo>        Use <viminfo> instead of .viminfo
+  -i <nviminfo>         Use <nviminfo> instead of .nviminfo
    --api-info      Dump API metadata serialized to msgpack and exit
-   --embed     Use stdin/stdout as a msgpack-rpc channel. This can be used for embedding Neovim into other programs
-   -h  or  --help  Print Help (this message) and exit
+  --embed               Use stdin/stdout as a msgpack-rpc channel
    --version       Print version information and exit
+  -h | --help           Print this help message and exit
```

### CODE REVIEW

**Comment by Pyrohh** [comment]
I realized the diff shows a small issue, where I accidentally removed the `[arguments]` fields. I'll fix that when I'm at a computer.

**Comment by Pyrohh** [comment]
Repushed and updated the diff.

**Comment by elmart** [comment]
I like the increase in clarity these changes bring. Note, though, that this breaks all existing translations for usage message. I think it's a price we can pay now. But it's something that shouldn't go unnoticed.

**Comment by Pyrohh** [comment]
Removed `puts` usage in favor of `mch_msg` (until we fully understand it) and rebased/pushed.

**Comment by elmart** [comment]
LGTM.

**Comment by Pyrohh** [comment]
Any other feedback, or may I mark this as RDY?

**Comment by elmart** [comment]
I suppose you mean feedback from other people. 
For what respects to me, it's ready.

File: src/nvim/main.c [hunk 1]
```diff
@@ -2057,93 +2057,66 @@ static void mainerr(int n, const char *str)
...
-{
-  mch_msg("   ");
-  mch_msg(s);
-  mch_msg("\n");
-}
-
-/*
- * Print messages for "vim -h" or "vim --help" and exit.
- */
+/// Prints help message and exits
```
Comment (justinmk) [inline src/nvim/main.c]: Noting "vim -h" or "vim --help" is useful, especially for grepping, but also for zero-friction comprehension of what a function does.

Comment (Pyrohh) [inline src/nvim/main.c]: Noted, I hadn't considered grepping.

**Comment by Pyrohh** [comment]
Rebased with the mentioned change.

> I suppose you mean feedback from other people.

Yes, I wasn't clear.

**Comment by Pyrohh** [comment]
Marking as ready.

File: src/nvim/main.c [hunk 2]
```diff
@@ -2057,93 +2057,66 @@ static void mainerr(int n, const char *str)
...
-  {
-    N_("[file ..]       edit specified file(s)"),
-    N_("-               read text from stdin"),
-    N_("-t tag          edit file where tag is defined"),
-    N_("-q [errorfile]  edit file with first error")
-  };
-
-  signal_stop();              /* kill us with CTRL-C here, if you like */
-
-  mch_msg(longVersion);
```
Comment (justinmk) [inline src/nvim/main.c]: Why not print `longVersion`? It's a pretty common tradition with unix tools.

File: src/nvim/main.c [hunk 3]
```diff
@@ -2057,93 +2057,66 @@ static void mainerr(int n, const char *str)
...
+  mch_msg(_("  -m                    Modifications (writing files) not allowed\n"));
+  mch_msg(_("  -M                    Modifications in text not allowed\n"));
+  mch_msg(_("  -b                    Binary mode\n"));
+  mch_msg(_("  -l                    Lisp mode\n"));
+  mch_msg(_("  -V[N][file]           Be verbose [level N][log messages to file]\n"));
+  mch_msg(_("  -D                    Debugging mode\n"));
+  mch_msg(_("  -n                    No swap file, use memory only\n"));
+  mch_msg(_("  -r                    List swap files and exit\n"));
+  mch_msg(_("  -r <file>             Recover crashed session\n"));
+  mch_msg(_("  -L                    Same as -r\n"));
```
Comment (justinmk) [inline src/nvim/main.c]: `-L` appears to be a vi equivalent. I don't know why `-r` exists also, but since `-L` does the same thing, there is zero reason to mention it in the already-very-busy `Usage:` message. Of course, we will still continue to support `-L` and document it in `:help`. 

But for `-h` message, eliminating noise is a high priority.

File: src/nvim/main.c [hunk 4]
```diff
@@ -2057,93 +2057,66 @@ static void mainerr(int n, const char *str)
...
+  mch_msg(_("  -H                    Start in Hebrew mode\n"));
+  mch_msg(_("  -T <terminal>         Set terminal type to <terminal>\n"));
+  mch_msg(_("  -u <nvimrc>           Use <nvimrc> instead of any .nvimrc\n"));
+  mch_msg(_("  --noplugin            Don't load plugin scripts\n"));
+  mch_msg(_("  -p[N]                 Open N tab pages (default: one for each file)\n"));
+  mch_msg(_("  -o[N]                 Open N windows (default: one for each file)\n"));
+  mch_msg(_("  -O[N]                 Like -o but split vertically\n"));
+  mch_msg(_("  +                     Start at end of file\n"));
+  mch_msg(_("  +<lnum>               Start at line <lnum>\n"));
+  mch_msg(_("  --cmd <command>       Execute <command> before loading any vimrc file\n"));
```
Comment (justinmk) [inline src/nvim/main.c]: nvimrc

File: src/nvim/main.c (hunk 2 above)
Comment (Pyrohh) [inline src/nvim/main.c]: I haven't seen it done anywhere else, but regardless I think it's just noise. I don't see how it's in any way helpful given that --version and the intro screen already include it.

File: src/nvim/main.c (hunk 4 above)
Comment (Pyrohh) [inline src/nvim/main.c]: :+1:

File: src/nvim/main.c (hunk 3 above)
Comment (Pyrohh) [inline src/nvim/main.c]: > Of course, we will still continue to support -L and document it in :help

IMO, dealing with too many arguments via selectively removing them from the help dialogue while still providing documentation is just avoiding the issue.

Why not just remove the ability to use `-r` without arguments, as `-L` is easier to remember. Even better, just remove lisp mode so that `-l` can be used instead. Just to be clear, lisp mode is described as such:
```
-l              Lisp mode.  Sets the 'lisp' and 'showmatch' options on.
```
`lisp` is already set via `runtime/ftplugin/lisp.vim`, and `showmatch` could just be set by the user, whether it be via a local ftplugin or globally.

I could also get rid of `-v`, and just rename this PR & commits to something like `Misc. interface improvements`, but that's up to you.

Comment (justinmk) [inline src/nvim/main.c]: > IMO, dealing with too many arguments via selectively removing them from the help dialogue while still providing documentation is just avoiding the issue.

Avoiding the issue is exactly what we want right now. Removing a redundant option from the "Usage" output is *zero* cost, plus a small net benefit. 

> Why not just remove the ability ....

Because it could break things and it's a low priority with a non-zero cost :)

Comment (Pyrohh) [inline src/nvim/main.c]: > Avoiding the issue is exactly what we want right now

That seems reasonable, I'll remove it (for now :smile:).

**Comment by Pyrohh** [comment]
Repushed

**Comment by Pyrohh** [comment]
Repushed.

**Comment by justinmk** [comment]
LGTM after squash

**Comment by Pyrohh** [comment]
Sounds good, but just one question: wouldn't it be easier for you if you pulled my branch, squash my commits, and then cherry picked onto master instead? Using `hub` makes this trivial; just `hub checkout [this PRs URL]` and it will add my repo as a remote, fetch it, and checkout into it. It's very useful for reviews as well.

I'm not trying to avoid work, just saying that if I squashed my commits it would create another commit on GH for you to review, presumably one you already have (as something could have been snuck in the commit under the guise of a rebase). If you pulled my branch and squashed yourself, then you wouldn't need to re-review the commits (right..?), although then again this really just offloads the trust from me to you. Additionally, you'd avoid a merge commit for only one commit, which I've noticed you've been doing recently.

I'm not trying to start some big debate, so feel free to not take this to heart. Just wanted to put it out there.

**Comment by justinmk** [comment]
In this case, sure, I asked out of habit, and it was no problem to squash. But in general, authors should reorder/squash/organize their own commits, otherwise we'll end up with a bottleneck situation same as Vim, where Bram does all that stuff himself.

**Comment by justinmk** [comment]
Also, rebasing and repushing is a good habit for us to follow because it can have unexpected consequences that should be re-checked by travis.

**Comment by Pyrohh** [comment]
I understand, I wouldn't want such a bottleneck situation.
